src/job-cache/job-cache.cpp

```diff
@@ -0,0 +1,1033 @@
+  }
+  return out;
+}
+
+static std::vector<std::string> split_path(const std::string &path) {
+  std::vector<std::string> path_vec;
+  for (std::string node : wcl::make_filepath_range_ref(path)) {
+    path_vec.emplace_back(std::move(node));
+  }
+
+  return path_vec;
+}
+
+// The very last value is assumed to be a file and
+// is not created
+template <class Iter>
+static void mkdir_all(Iter begin, Iter end) {
+  std::string acc;
+  for (; begin + 1 != end; ++begin) {
+    acc += *begin + "/";
+    mkdir_no_fail(acc.c_str());
+  }
+}
+
+// TODO: Do all file movement in parallel
+
+}  // namespace
+
+namespace job_cache {
+
+struct CacheDbImpl {
+ private:
+  std::shared_ptr<Database> db;
+
+ public:
+  JobTable jobs;
+  InputFiles input_files;
+  InputDirs input_dirs;
+  OutputFiles output_files;
+  Transaction transact;
+  SelectMatchingJobs matching_jobs;
+
+  CacheDbImpl(const std::string &_dir)
+      : db(std::make_unique<Database>(_dir)),
+        jobs(db),
+        input_files(db),
+        input_dirs(db),
+        output_files(db),
+        transact(db),
+        matching_jobs(db) {}
+};
+
+AddJobRequest::AddJobRequest(const JAST &job_result_json) {
+  cwd = job_result_json.get("cwd").value;
+  command_line = job_result_json.get("command_line").value;
+  envrionment = job_result_json.get("envrionment").value;
+  stdin_str = job_result_json.get("stdin").value;
+  stdout_str = job_result_json.get("stdout").value;
+  stderr_str = job_result_json.get("stderr").value;
+  ret_code = std::stoi(job_result_json.get("status").value);
+  runtime = std::stod(job_result_json.get("runtime").value);
+  cputime = std::stod(job_result_json.get("cputime").value);
+  mem = std::stoull(job_result_json.get("mem").value);
+  ibytes = std::stoull(job_result_json.get("ibytes").value);
+  obytes = std::stoull(job_result_json.get("obytes").value);
+
+  // Read the input files
+  for (const auto &input_file : job_result_json.get("input_files").children) {
+    InputFile input;
+    input.path = input_file.second.get("path").value;
+    input.hash = Hash256::from_hex(input_file.second.get("hash").value);
+    bloom.add_hash(input.hash);
+    inputs.emplace_back(std::move(input));
+  }
+
+  // Read the input dirs
+  for (const auto &input_dir : job_result_json.get("input_dirs").children) {
+    InputDir input;
+    input.path = input_dir.second.get("path").value;
+    input.hash = Hash256::from_hex(input_dir.second.get("hash").value);
+    bloom.add_hash(input.hash);
+    directories.emplace_back(std::move(input));
+  }
+
+  // TODO: I hate this loop but its the fastest path to a demo.
+  //       we need to figure out a path add things to the cache
+  //       only after the files have been hashed so we don't
+  //       need this loop.
+  std::vector<std::future<OutputFile>> future_outputs;
+
+  // Read the output files which requires kicking off a hash
+  for (const auto &output_file : job_result_json.get("output_files").children) {
```

Comment:
It's not part of the same schema so I'm not sure its really being reused.

Reply:
Let me know where you're landing on this issue still. I can still add constexpr fields for all these things but there's currently no reuse.

Reply:
It's not worth blocking on so feel free to keep it like this!

(not anyone fault nor something even related to this PR) I'm generally unhappy with how we do SQL stuff in wake but I don't have a good solution or alternative so until that point we should leave things as is. This is mostly just my back-end/API dev past self grumbling since I know a world where this is so much better (PHP surprisingly).